Build broken with MacOS Sonoma 14.5

```
➜  specter-diy git:(master) make disco                                                      
mkdir -p bin
git submodule update --init --recursive
Submodule 'bootloader' (https://github.com/cryptoadvance/specter-bootloader) registered for path 'bootloader'
Submodule 'f469-disco' (https://github.com/diybitcoinhardware/f469-disco) registered for path 'f469-disco'
Cloning into '/Users/kim/src/specter-diy/bootloader'...
Cloning into '/Users/kim/src/specter-diy/f469-disco'...
Submodule path 'bootloader': checked out 'fc6e61eeada07c36b00c99f18f74c26b95b36ddb'
Submodule 'lib/fatfs' (https://github.com/cryptoadvance/fatfs.git) registered for path 'bootloader/lib/fatfs'
Submodule 'lib/secp256k1' (https://github.com/bitcoin-core/secp256k1.git) registered for path 'bootloader/lib/secp256k1'
Cloning into '/Users/kim/src/specter-diy/bootloader/lib/fatfs'...
Cloning into '/Users/kim/src/specter-diy/bootloader/lib/secp256k1'...
Submodule path 'bootloader/lib/fatfs': checked out '8ea398023265f9fdec98b3a3f7c670681b1484cf'
Submodule path 'bootloader/lib/secp256k1': checked out '5e1c885efb0f400d024efc259eddd6a5ee9cba7b'
Submodule path 'f469-disco': checked out 'db3ce3e918cf0fd36f076ecd86ef05240d7c3cef'
Submodule 'micropython' (https://github.com/diybitcoinhardware/micropython) registered for path 'f469-disco/micropython'
Submodule 'usermods/secp256k1' (https://github.com/diybitcoinhardware/secp256k1-embedded) registered for path 'f469-disco/usermods/secp256k1'
Submodule 'usermods/udisplay_f469/lvgl' (https://github.com/littlevgl/lvgl) registered for path 'f469-disco/usermods/udisplay_f469/lvgl'
Cloning into '/Users/kim/src/specter-diy/f469-disco/micropython'...
Cloning into '/Users/kim/src/specter-diy/f469-disco/usermods/secp256k1'...
Cloning into '/Users/kim/src/specter-diy/f469-disco/usermods/udisplay_f469/lvgl'...
Submodule path 'f469-disco/micropython': checked out '6bdf1b69162b673d48042ccd021f9efa019091fa'
Submodule path 'f469-disco/usermods/secp256k1': checked out '1e74fc3c617222396d9a49dd53f6c4bb28a9c117'
Submodule 'secp256k1' (https://github.com/ElementsProject/secp256k1-zkp.git) registered for path 'f469-disco/usermods/secp256k1/secp256k1'
Cloning into '/Users/kim/src/specter-diy/f469-disco/usermods/secp256k1/secp256k1'...
Submodule path 'f469-disco/usermods/secp256k1/secp256k1': checked out 'd9560e0af78d9059bba0c4845a310387abfa4e5e'
Submodule path 'f469-disco/usermods/udisplay_f469/lvgl': checked out 'dd100e5e07c8ae18c6e885d97d9c5938049fff44'
Building cross-compiler
make -C f469-disco/micropython/mpy-cross \
        DEBUG=0 && \
        cp f469-disco/micropython/mpy-cross/mpy-cross bin
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
mkdir -p build/genhdr
GEN build/genhdr/mpversion.h
GEN build/genhdr/moduledefs.h
GEN build/genhdr/qstr.i.last
GEN build/genhdr/qstr.split
GEN build/genhdr/qstrdefs.collected.h
QSTR updated
GEN build/genhdr/qstrdefs.generated.h
mkdir -p build/py
CC ../py/mpstate.c
CC ../py/nlr.c
CC ../py/nlrx86.c
CC ../py/nlrx64.c
CC ../py/nlrthumb.c
CC ../py/nlrpowerpc.c
CC ../py/nlrxtensa.c
CC ../py/nlrsetjmp.c
CC ../py/malloc.c
CC ../py/gc.c
CC ../py/pystack.c
CC ../py/qstr.c
CC ../py/vstr.c
CC ../py/mpprint.c
CC ../py/unicode.c
CC ../py/mpz.c
CC ../py/reader.c
CC ../py/lexer.c
CC ../py/parse.c
CC ../py/scope.c
CC ../py/compile.c
CC ../py/emitcommon.c
CC ../py/emitbc.c
CC ../py/asmbase.c
CC ../py/asmx64.c
CC ../py/emitnx64.c
CC ../py/asmx86.c
CC ../py/emitnx86.c
CC ../py/asmthumb.c
CC ../py/emitnthumb.c
CC ../py/emitinlinethumb.c
CC ../py/asmarm.c
CC ../py/emitnarm.c
CC ../py/asmxtensa.c
CC ../py/emitnxtensa.c
CC ../py/emitinlinextensa.c
CC ../py/emitnxtensawin.c
CC ../py/formatfloat.c
CC ../py/parsenumbase.c
CC ../py/parsenum.c
CC ../py/emitglue.c
CC ../py/persistentcode.c
CC ../py/runtime.c
CC ../py/runtime_utils.c
CC ../py/scheduler.c
CC ../py/nativeglue.c
CC ../py/ringbuf.c
CC ../py/stackctrl.c
CC ../py/argcheck.c
CC ../py/warning.c
CC ../py/profile.c
CC ../py/map.c
CC ../py/obj.c
CC ../py/objarray.c
CC ../py/objattrtuple.c
CC ../py/objbool.c
CC ../py/objboundmeth.c
CC ../py/objcell.c
CC ../py/objclosure.c
CC ../py/objcomplex.c
CC ../py/objdeque.c
CC ../py/objdict.c
CC ../py/objenumerate.c
CC ../py/objexcept.c
CC ../py/objfilter.c
CC ../py/objfloat.c
CC ../py/objfun.c
../py/objfun.c:142:5: error: variable 'n_info' set but not used [-Werror,-Wunused-but-set-variable]
  142 |     MP_BC_PRELUDE_SIZE_DECODE(code_info);
      |     ^
../py/bc.h:165:12: note: expanded from macro 'MP_BC_PRELUDE_SIZE_DECODE'
  165 |     size_t n_info, n_cell; \
      |            ^
../py/objfun.c:142:5: error: variable 'n_cell' set but not used [-Werror,-Wunused-but-set-variable]
../py/bc.h:165:20: note: expanded from macro 'MP_BC_PRELUDE_SIZE_DECODE'
  165 |     size_t n_info, n_cell; \
      |                    ^
../py/objfun.c:164:5: error: variable 'n_state' set but not used [-Werror,-Wunused-but-set-variable]
  164 |     MP_BC_PRELUDE_SIG_DECODE(bc);
      |     ^
../py/bc.h:130:12: note: expanded from macro 'MP_BC_PRELUDE_SIG_DECODE'
  130 |     size_t n_state, n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_pos_args; \
      |            ^
../py/objfun.c:164:5: error: variable 'n_exc_stack' set but not used [-Werror,-Wunused-but-set-variable]
../py/bc.h:130:21: note: expanded from macro 'MP_BC_PRELUDE_SIG_DECODE'
  130 |     size_t n_state, n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_pos_args; \
      |                     ^
../py/objfun.c:164:5: error: variable 'scope_flags' set but not used [-Werror,-Wunused-but-set-variable]
../py/bc.h:130:34: note: expanded from macro 'MP_BC_PRELUDE_SIG_DECODE'
  130 |     size_t n_state, n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_pos_args; \
      |                                  ^
../py/objfun.c:164:5: error: variable 'n_pos_args' set but not used [-Werror,-Wunused-but-set-variable]
../py/bc.h:130:47: note: expanded from macro 'MP_BC_PRELUDE_SIG_DECODE'
  130 |     size_t n_state, n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_pos_args; \
      |                                               ^
../py/objfun.c:164:5: error: variable 'n_kwonly_args' set but not used [-Werror,-Wunused-but-set-variable]
../py/bc.h:130:59: note: expanded from macro 'MP_BC_PRELUDE_SIG_DECODE'
  130 |     size_t n_state, n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_pos_args; \
      |                                                           ^
../py/objfun.c:164:5: error: variable 'n_def_pos_args' set but not used [-Werror,-Wunused-but-set-variable]
../py/bc.h:130:74: note: expanded from macro 'MP_BC_PRELUDE_SIG_DECODE'
  130 |     size_t n_state, n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_pos_args; \
      |                                                                          ^
../py/objfun.c:255:5: error: variable 'scope_flags' set but not used [-Werror,-Wunused-but-set-variable]
  255 |     DECODE_CODESTATE_SIZE(self->bytecode, n_state, state_size);
      |     ^
../py/objfun.c:196:29: note: expanded from macro 'DECODE_CODESTATE_SIZE'
  196 |         size_t n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_args; \
      |                             ^
../py/objfun.c:255:5: error: variable 'n_pos_args' set but not used [-Werror,-Wunused-but-set-variable]
../py/objfun.c:196:42: note: expanded from macro 'DECODE_CODESTATE_SIZE'
  196 |         size_t n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_args; \
      |                                          ^
../py/objfun.c:255:5: error: variable 'n_kwonly_args' set but not used [-Werror,-Wunused-but-set-variable]
../py/objfun.c:196:54: note: expanded from macro 'DECODE_CODESTATE_SIZE'
  196 |         size_t n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_args; \
      |                                                      ^
../py/objfun.c:255:5: error: variable 'n_def_args' set but not used [-Werror,-Wunused-but-set-variable]
../py/objfun.c:196:69: note: expanded from macro 'DECODE_CODESTATE_SIZE'
  196 |         size_t n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_args; \
      |                                                                     ^
12 errors generated.
make[1]: *** [build/py/objfun.o] Error 1
make: *** [mpy-cross] Error 2
➜  specter-diy git:(master)
```

Here is the version of the compiler:

```
➜  bin brew install arm-none-eabi-gcc
==> Auto-updating Homebrew...
Adjust how often this is run with HOMEBREW_AUTO_UPDATE_SECS or disable with
HOMEBREW_NO_AUTO_UPDATE. Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
==> Auto-updated Homebrew!
Updated 2 taps (homebrew/core and homebrew/cask).
==> New Formulae
fortitude     gersemi       lazyjournal   mox           rhai          rpds-py       sequoia-sqv   typioca       unciv
==> New Casks
focu                             macskk                           nvidia-nsight-compute            trae-cn

You have 8 outdated formulae and 1 outdated cask installed.

Warning: arm-none-eabi-gcc 14.2.0 is already installed and up-to-date.
To reinstall 14.2.0, run:
  brew reinstall arm-none-eabi-gcc
➜  bin 
```

If i ignore those warnings, i get another bunch of errors:
```
➜  specter-diy git:(v1.9.0) make CFLAGS_EXTRA="-Wno-unused-but-set-variable" disco
mkdir -p bin
Building cross-compiler
make -C f469-disco/micropython/mpy-cross \
        DEBUG=0 && \
        cp f469-disco/micropython/mpy-cross/mpy-cross bin
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
mkdir -p build/genhdr
GEN build/genhdr/mpversion.h
GEN build/genhdr/moduledefs.h
GEN build/genhdr/qstr.i.last
GEN build/genhdr/qstr.split
GEN build/genhdr/qstrdefs.collected.h
QSTR updated
GEN build/genhdr/qstrdefs.generated.h
mkdir -p build/py
CC ../py/mpstate.c
CC ../py/nlr.c
CC ../py/nlrx86.c
CC ../py/nlrx64.c
CC ../py/nlrthumb.c
CC ../py/nlrpowerpc.c
CC ../py/nlrxtensa.c
CC ../py/nlrsetjmp.c
CC ../py/malloc.c
CC ../py/gc.c
CC ../py/pystack.c
CC ../py/qstr.c
CC ../py/vstr.c
CC ../py/mpprint.c
CC ../py/unicode.c
CC ../py/mpz.c
CC ../py/reader.c
CC ../py/lexer.c
CC ../py/parse.c
CC ../py/scope.c
CC ../py/compile.c
CC ../py/emitcommon.c
CC ../py/emitbc.c
CC ../py/asmbase.c
CC ../py/asmx64.c
CC ../py/emitnx64.c
CC ../py/asmx86.c
CC ../py/emitnx86.c
CC ../py/asmthumb.c
CC ../py/emitnthumb.c
CC ../py/emitinlinethumb.c
CC ../py/asmarm.c
CC ../py/emitnarm.c
CC ../py/asmxtensa.c
CC ../py/emitnxtensa.c
CC ../py/emitinlinextensa.c
CC ../py/emitnxtensawin.c
CC ../py/formatfloat.c
CC ../py/parsenumbase.c
CC ../py/parsenum.c
CC ../py/emitglue.c
CC ../py/persistentcode.c
CC ../py/runtime.c
CC ../py/runtime_utils.c
CC ../py/scheduler.c
CC ../py/nativeglue.c
CC ../py/ringbuf.c
CC ../py/stackctrl.c
CC ../py/argcheck.c
CC ../py/warning.c
CC ../py/profile.c
CC ../py/map.c
CC ../py/obj.c
CC ../py/objarray.c
CC ../py/objattrtuple.c
CC ../py/objbool.c
CC ../py/objboundmeth.c
CC ../py/objcell.c
CC ../py/objclosure.c
CC ../py/objcomplex.c
CC ../py/objdeque.c
CC ../py/objdict.c
CC ../py/objenumerate.c
CC ../py/objexcept.c
CC ../py/objfilter.c
CC ../py/objfloat.c
CC ../py/objfun.c
CC ../py/objgenerator.c
CC ../py/objgetitemiter.c
CC ../py/objint.c
CC ../py/objint_longlong.c
CC ../py/objint_mpz.c
CC ../py/objlist.c
CC ../py/objmap.c
CC ../py/objmodule.c
CC ../py/objobject.c
CC ../py/objpolyiter.c
CC ../py/objproperty.c
CC ../py/objnone.c
CC ../py/objnamedtuple.c
CC ../py/objrange.c
CC ../py/objreversed.c
CC ../py/objset.c
CC ../py/objsingleton.c
CC ../py/objslice.c
CC ../py/objstr.c
CC ../py/objstrunicode.c
CC ../py/objstringio.c
CC ../py/objtuple.c
CC ../py/objtype.c
CC ../py/objzip.c
CC ../py/opmethods.c
CC ../py/sequence.c
CC ../py/stream.c
CC ../py/binary.c
CC ../py/builtinimport.c
CC ../py/builtinevex.c
CC ../py/builtinhelp.c
CC ../py/modarray.c
CC ../py/modbuiltins.c
CC ../py/modcollections.c
CC ../py/modgc.c
CC ../py/modio.c
CC ../py/modmath.c
CC ../py/modcmath.c
CC ../py/modmicropython.c
CC ../py/modstruct.c
CC ../py/modsys.c
CC ../py/moduerrno.c
CC ../py/modthread.c
CC ../py/vm.c
../py/vm.c:805:25: error: array index -3 refers past the last possible element for an array in 64-bit address space containing 64-bit (8-byte) elements (max possible 2305843009213693952 elements) [-Werror,-Warray-bounds]
  805 |                         sp[-MP_OBJ_ITER_BUF_NSLOTS + 1] = MP_OBJ_NULL;
      |                         ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
../py/vm.c:267:13: note: array 'sp' declared here
  267 |             mp_obj_t *sp = code_state->sp;
      |             ^
../py/vm.c:806:25: error: array index -2 refers past the last possible element for an array in 64-bit address space containing 64-bit (8-byte) elements (max possible 2305843009213693952 elements) [-Werror,-Warray-bounds]
  806 |                         sp[-MP_OBJ_ITER_BUF_NSLOTS + 2] = obj;
      |                         ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
../py/vm.c:267:13: note: array 'sp' declared here
  267 |             mp_obj_t *sp = code_state->sp;
      |             ^
../py/vm.c:817:25: error: array index -3 refers past the last possible element for an array in 64-bit address space containing 64-bit (8-byte) elements (max possible 2305843009213693952 elements) [-Werror,-Warray-bounds]
  817 |                     if (sp[-MP_OBJ_ITER_BUF_NSLOTS + 1] == MP_OBJ_NULL) {
      |                         ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
../py/vm.c:267:13: note: array 'sp' declared here
  267 |             mp_obj_t *sp = code_state->sp;
      |             ^
../py/vm.c:818:31: error: array index -2 refers past the last possible element for an array in 64-bit address space containing 64-bit (8-byte) elements (max possible 2305843009213693952 elements) [-Werror,-Warray-bounds]
  818 |                         obj = sp[-MP_OBJ_ITER_BUF_NSLOTS + 2];
      |                               ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
../py/vm.c:267:13: note: array 'sp' declared here
  267 |             mp_obj_t *sp = code_state->sp;
      |             ^
4 errors generated.
make[1]: *** [build/py/vm.o] Error 1
make: *** [mpy-cross] Error 2
➜  specter-diy git:(v1.9.0) 
````

This seems to also happen if i install this [gcc-arm-none-eabi-9-2020-q2-update-mac.pkg](https://developer.arm.com/-/media/Files/downloads/gnu-rm/9-2020q2/gcc-arm-none-eabi-9-2020-q2-update-mac.pkg?revision=05bece2b-0d55-4247-a46c-25d6539c5a6e&rev=05bece2b0d554247a46c25d6539c5a6e&hash=7020405F78DCFE79EE6B1A5645C0CE8B5149751C) from this [page](https://developer.arm.com/downloads/-/gnu-rm/9-2020-q2-update).

--- a/.envrc
+++ b/.envrc
@@ -1,0 +1,1 @@
+use flake

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -1,0 +1,43 @@
+name: Build
+
+on:
+  pull_request:
+    branches: [main, master]
+  push:
+    branches: [main, master]
+
+jobs:
+  build:
+    name: Build Firmware
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      
+      - name: Install Nix
+        uses: DeterminateSystems/nix-installer-action@v14
+
+      - name: Cache Nix store
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cache/nix
+            /nix/store
+          key: nix-${{ runner.os }}-${{ hashFiles('flake.lock') }}
+          restore-keys: |
+            nix-${{ runner.os }}-
+
+      - name: Build Unix simulator
+        run: nix develop -c make unix
+
+      - name: Build STM32F469 Discovery firmware
+        run: nix develop -c make disco
+
+      - name: Upload firmware artifacts
+        uses: actions/upload-artifact@v4
+        with:
+          name: firmware-binaries
+          path: |
+            bin/specter-diy.bin
+            bin/specter-diy.hex

--- a/.gitignore
+++ b/.gitignore
@@ -8,3 +8,4 @@ release
 .idea
 .vscode
 .DS_Store
+.direnv

--- a/docs/build.md
+++ b/docs/build.md
@@ -6,11 +6,35 @@ Clone the repository recursively `git clone https://github.com/cryptoadvance/spe
 
 ## Prerequisities
 
+There are multiple ways to get all necessary tools. The recommended way is to use the Nix flake with direnv.
+If that's too complicated for you, you can use the traditional `nix-shell` or install the tools manually (which mighty be tricky to get the dependencies right).
+
+### Nix flake (Recommended)
+
+The easiest way to get all necessary tools is to use the Nix flake from the root of the repository. You need to have [Nix](https://nixos.org/) (on Mac use [determinate](https://github.com/DeterminateSystems/nix-installer)) with flakes enabled.
+Install direnv with `brew install direnv` (on Mac) or `sudo apt install direnv` (on Linux).
+
+Make sure that [flakes are enabled](https://nixos.wiki/wiki/Flakes) in your Nix config.
+
+```sh
+# Enter development shell
+nix develop
+
+# Or use with direnv for automatic activation
+direnv allow
+```
+
+
 ### Nix shell
 
-The easiest way to get all necessary tools is to run `nix-shell` from the root of the repository. You need to have [Nix](https://nixos.org/) installed.
+Alternatively, you can use the traditional `shell.nix`:
 
-### Prerequisities: Board
+```sh
+nix-shell
+```
+You'll need to have [Nix](https://nixos.org/) installed as well.
+
+### Prerequisities (Manually): Board
 
 To compile the firmware for the board you will need `arm-none-eabi-gcc` compiler.
 
@@ -34,7 +58,7 @@ brew install arm-none-eabi-gcc
 
 On **Windows**: Install linux subsystem and follow Linux instructions.
 
-### Prerequisities: Simulator
+### Prerequisities (Manually): Simulator
 
 You may need to install SDL2 library to simulate the screen of the device.
 

--- a/docs/build.md
+++ b/docs/build.md
@@ -11,10 +11,10 @@ If that's too complicated for you, you can use the traditional `nix-shell` or in
 
 ### Nix flake (Recommended)
 
-The easiest way to get all necessary tools is to use the Nix flake from the root of the repository. You need to have [Nix](https://nixos.org/) (on Mac use [determinate](https://github.com/DeterminateSystems/nix-installer)) with flakes enabled.
+The easiest way to get all necessary tools is to use the Nix flake from the root of the repository. You need to have [Nix](https://nixos.org/) (on Mac use [determinate](https://github.com/DeterminateSystems/nix-installer)) with flakes enabled. This only works with Nix >2.7 (check with `nix --version`).
 Install direnv with `brew install direnv` (on Mac) or `sudo apt install direnv` (on Linux).
 
-Make sure that [flakes are enabled](https://nixos.wiki/wiki/Flakes) in your Nix config.
+Make sure that [flakes are enabled](https://nixos.wiki/wiki/Flakes) in your Nix config. On Linux systems, your user might need to be added to the nix-users group.
 
 ```sh
 # Enter development shell

--- a/docs/build.md
+++ b/docs/build.md
@@ -79,6 +79,8 @@ brew install sdl2
 
 ## Build
 
+All build commands might need a prefix like `nix develop -c` or need to be run from `nix develop` shell. If you use direnv, you don't need to do anything, apart from an initial `direnv allow`.
+
 To build custom bootloader and firmware that you will be able to sign check out the bootloader doc on [self-signed firmware](https://github.com/cryptoadvance/specter-bootloader/blob/master/doc/selfsigned.md). To wipe flash and remove protections on the device with the secure bootloader check out [this doc](https://github.com/cryptoadvance/specter-bootloader/blob/master/doc/remove_protection.md).
 
 To build an open firmware (no bootloader and signature verifications) run `make disco`. The result is the `bin/specter-diy.bin` file that you can copy-paste to the board over miniUSB.

--- a/docs/development.md
+++ b/docs/development.md
@@ -2,29 +2,36 @@
 
 ## Compiling the code yourself
 
-We use this build as a platform for Specter: https://github.com/diybitcoinhardware/f469-disco
-
-To compile the firmware you will need `arm-none-eabi-gcc` compiler.
-
-On MacOS install it using brew: `brew install arm-none-eabi-gcc`
-
-On Debian: `sudo apt install gcc-arm-none-eabi binutils-arm-none-eabi gdb-arm-none-eabi openocd`
-
-On Arch Linux: `sudo pacman -S arm-none-eabi-gcc arm-none-eabi-binutils arm-none-eabi-gdb arm-none-eabi-newlib openocd`
-
-Run `make disco` to get the binary or `make unix` to compile the simulator. They will be in the `bin` folder.
-
-`specter-diy.bin` file is the firmware that you need to copy to the device.
-
-The easiest way to start developing is to use a [simulator](./simulator.md), and when you are done - try it on a real hardware.
+See the [build documentation](build.md) for instructions on how to compile the code.
 
 ## Enabling developer mode
 
 By default developer mode and USB communication are turned off. This means that when you connect the board to the computer it will NOT mount the `PYBFLASH` anymore and there will be no way to connect to debug shell.
 
-To turn on the developer mode get to the main screen (enter PIN code, generate recovery phrase, enter password), and then go to **Settings - Security - turn on Developer mode - Save**.
 
-Now the board will restart and get mounted to the computer as before. You can also connect to the board over miniUSB and get to interactive console (baudrate 115200). You can use `screen` or `putty` or `minicom` for that, i.e. `screen /dev/tty.usbmodem14403 115200`.
+~~To turn on the developer mode get to the main screen (enter PIN code, generate recovery phrase, enter password), and then go to **Settings - Security - turn on Developer mode - Save**.~~
+(Currently deactivated for securtiy reasons)
+
+~~Now the board will restart and get mounted to the computer as before. You can also connect to the board over miniUSB and get to interactive console (baudrate 115200). You can use `screen` or `putty` or `minicom` for that, i.e. `screen /dev/tty.usbmodem14403 115200`.~~
+
+In order to connect to the board, modify those lines in [boot.py](https://github.com/cryptoadvance/specter-diy/blob/2f51e152bcdb184cf719792e6c5f972214e3dd36/boot/main/boot.py#L33-L40) like this:
+```python
+# configure usb from start if you want, 
+# otherwise will be configured after PIN
+pyb.usb_mode("VCP+MSC") # debug mode with USB and mounted storages from start
+#pyb.usb_mode("VCP") # debug mode with USB from start
+# disable at start
+# pyb.usb_mode(None)
+```
+
+and after flashing, you can connect with something like:
+```
+# Linux
+screen /dev/ttyACM0 115200 # or maybe ttyACM1 or ttyACM2
+# Mac
+screen /dev/tty.usbmodem14403 115200`.
+```
+
 
 ## Writing a simple app
 

--- a/flake.nix
+++ b/flake.nix
@@ -1,0 +1,26 @@
+{
+  description = "Specter DIY development environment";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.05";
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { self, nixpkgs, flake-utils }:
+    flake-utils.lib.eachDefaultSystem (system:
+      let
+        pkgs = nixpkgs.legacyPackages.${system};
+      in
+      {
+        devShells.default = pkgs.mkShell {
+          nativeBuildInputs = [ 
+            pkgs.buildPackages.gcc-arm-embedded-9
+            pkgs.buildPackages.python39
+            pkgs.openocd
+            pkgs.stlink
+            pkgs.SDL2
+          ];
+          hardeningDisable = ["all"];
+        };
+      });
+}

--- a/flake.nix
+++ b/flake.nix
@@ -4,6 +4,10 @@
   inputs = {
     nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.05";
     flake-utils.url = "github:numtide/flake-utils";
+    flake-compat = {
+      url = "github:edolstra/flake-compat";
+      flake = false;
+    };
   };
 
   outputs = { self, nixpkgs, flake-utils }:

--- a/flake.nix
+++ b/flake.nix
@@ -10,7 +10,7 @@
     };
   };
 
-  outputs = { self, nixpkgs, flake-utils }:
+  outputs = { self, nixpkgs, flake-utils,  ... }:
     flake-utils.lib.eachDefaultSystem (system:
       let
         pkgs = nixpkgs.legacyPackages.${system};

--- a/shell.nix
+++ b/shell.nix
@@ -1,4 +1,8 @@
-{ pkgs ? import <nixpkgs> {} }:
+{ pkgs ? import (fetchTarball {
+    url = "https://github.com/NixOS/nixpkgs/archive/nixos-22.05.tar.gz";
+    sha256 = "154x9swf494mqwi4z8nbq2f0sp8pwp4fvx51lqzindjfbb9yxxv5";
+  }) {}
+}:
   pkgs.mkShell {
     nativeBuildInputs = [ 
       pkgs.buildPackages.gcc-arm-embedded-9

--- a/shell.nix
+++ b/shell.nix
@@ -1,4 +1,19 @@
-{ pkgs ? import (fetchTarball {
+(import
+  (
+    let
+      lock = builtins.fromJSON (builtins.readFile ./flake.lock);
+    in
+    fetchTarball {
+      url =
+        lock.nodes.flake-compat.locked.url
+          or "https://github.com/edolstra/flake-compat/archive/${lock.nodes.flake-compat.locked.rev}.tar.gz";
+      sha256 = lock.nodes.flake-compat.locked.narHash;
+    }
+  )
+  {
+    src = ./.;
+  }
+).shellNix
     url = "https://github.com/NixOS/nixpkgs/archive/nixos-22.05.tar.gz";
     sha256 = "154x9swf494mqwi4z8nbq2f0sp8pwp4fvx51lqzindjfbb9yxxv5";
   }) {}

